lib/Models/RegionProvider.js

```diff
@@ -0,0 +1,405 @@
+    }            
+    var that = this;
+
+    var baseuri = URI(this.server).addQuery({ 
+        service: 'wfs',
+        version: '2.0',
+        request: 'getPropertyValue',
+        typenames: this.layerName});
+
+    // get the list of IDs that we will attempt to match against for this column
+    var p = [];
+    var url = baseuri.addSearch('valueReference', this.regionProp).toString();
+    p.push(loadText(corsProxy.getURL(url))
+        .then(function(xml) {
+            that.loadRegionsFromXML(xml);
+        }).otherwise(function(err) {
+            console.log(err);
+            throw(err);
+        }));
+    // if this column might be ambiguous then fetch the disambiguating values for each column as well (usually State)
+    var dab = this.disambigColumns;
+    if (defined(dab)) {
+        
+        url = baseuri.addQuery('valueReference' + dab[0].regionProp).toString();
+        p.push(loadText(corsProxy.getURL(url)).then(function(xml) {
+            that.loadRegionsFromXML(xml, dab[0].regionProp);
+        }));
+    }
+    return when.all(p).yield(true);
+};
+
+
+function applyReplacements(s, replacements) {
+// replacements: [ [ regex, replacement], ... ] 
+// replacement can contain '$1' etc.
+    var r = s.trim().toLowerCase(); // just in case...
+
+    if (replacements === undefined || replacements.length === 0) {
+        return r;
+    }
+
+    var cachekey = s + '%' + replacements.reduce(function(key, item) { 
+        return key + item[0] + '%' + (item[1] ? item[1] : '') + '%'; 
+        }, '');
+    if (applyReplacements.cache[cachekey] !== undefined) {
+        return applyReplacements.cache[cachekey];
+    }
+    
+    replacements.forEach(function(rep) {
+        r = r.replace(new RegExp(rep[0].toLowerCase(), 'gi'), rep[1].toLowerCase());
+    });
+    applyReplacements.cache[s] =r;
+    return r;
+}
+applyReplacements.cache = {};
+
+
+/* Checks if a given region code matches a given ID after applying regexs and additional disambiguation lookups */
+RegionProvider.prototype.codeMatchesRegionID = function(id, code, disambigId, disambigCode) { //, row, idnum, csvItem ) {
+    code = String(code).toLowerCase(); //## should be happening upstream somewhere
+    id = String(id).toLowerCase(); 
+    if (applyReplacements(id, this.serverReplacements) !== applyReplacements(code, this.dataReplacements)) {
+        return false; // failed match, even after best efforts
+    }
+    if (!defined(this.disambigColumns) || this.disambigColumns.length === 0) {
+        return true; // unambiguous successful match
+    }
+
+    if (defined(disambigId) && !defined(disambigCode)) {
+        // warn? we have a value to disambiguate against, but it's not provided
+        return true;
+    }
+
+    // Hooray, let's disambiguate. Our args look something like: Campbelltown, Campbelltown, New South Wales, NSW.
+
+    var dabMatched = this._getDisambigProvider().codeMatchesRegionID(disambigId, disambigCode);
```

Comment:
It seems a bit awkward for each region provider to keep a reference back to the list of providers that contains it just so that it can go find its disambiguation provider.

It would be better, I think, if the `regionProviderList` were passed down as parameters to the functions along the way, rather than given to the `RegionProvider` constructor and stored.

Also, I suspect that looking up the disambig provider with a a linear search on every attempted match is making the matching go quite a bit slower than it needs to.